sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AadAuthenticationProperties.java

```diff
@@ -552,100 +568,115 @@ private void validateApplicationType() {
-                        + "'authorization_code'.", registrationId);
-                    break;
-                case RESOURCE_SERVER:
-                case RESOURCE_SERVER_WITH_OBO:
-                    properties.setAuthorizationGrantType(AadAuthorizationGrantType.ON_BEHALF_OF);
-                    LOGGER.debug("The client '{}' sets the default value of AADAuthorizationGrantType to "
-                        + "'on_behalf_of'.", registrationId);
-                    break;
-                case WEB_APPLICATION_AND_RESOURCE_SERVER:
-                    throw new IllegalStateException("spring.cloud.azure.active-directory.authorization-clients." + registrationId
-                        + ".authorization-grant-grantType must be configured. ");
-                default:
-                    throw new IllegalStateException("Unsupported authorization grantType " + applicationType.getValue());
-            }
-        } else {
-            // Validate authorization grant grantType
-            switch (applicationType) {
-                case WEB_APPLICATION:
-                    if (ON_BEHALF_OF.getValue().equals(grantType)) {
-                        throw new IllegalStateException(String.format(UNMATCHING_OAUTH_GRANT_TYPE_FROMAT,
-                            "web_application", registrationId, "on_behalf_of"));
-                    }
-                    break;
-                case RESOURCE_SERVER:
-                    if (AUTHORIZATION_CODE.getValue().equals(grantType)) {
-                        throw new IllegalStateException(String.format(UNMATCHING_OAUTH_GRANT_TYPE_FROMAT,
-                            "resource_server", registrationId, "authorization_code"));
-                    }
-                    if (ON_BEHALF_OF.getValue().equals(grantType)) {
-                        throw new IllegalStateException(String.format(UNMATCHING_OAUTH_GRANT_TYPE_FROMAT,
-                            "resource_server", registrationId, "on_behalf_of"));
-                    }
-                    break;
-                case RESOURCE_SERVER_WITH_OBO:
-                    if (AUTHORIZATION_CODE.getValue().equals(grantType)) {
-                        throw new IllegalStateException(String.format(UNMATCHING_OAUTH_GRANT_TYPE_FROMAT,
-                            "resource_server_with_obo", registrationId, "authorization_code"));
-                    }
-                    break;
-                case WEB_APPLICATION_AND_RESOURCE_SERVER:
-                default:
-                    LOGGER.debug("'spring.cloud.azure.active-directory.authorization-clients.{}.authorization-grant-type' is valid.", registrationId);
-            }
+        if (grantType == null) {
+            grantType = decideDefaultGrantTypeFromApplicationType(registrationId, applicationType);
+            properties.setAuthorizationGrantType(grantType);
+        }
+
+        // Validate authorization grant grantType
+        validateAuthorizationGrantType(registrationId, grantType, properties);
```

Comment:
this should go to the else of `if (grantType == null)`

Reply:
This is intended to remove the else statement. For I find this mean to set the default value if null. Not if...else... logic.

Reply:
the validateAuthorizationGrantType is to validate input value from users, not for the value we set.

Reply:
well, but do an extra validate process will do no harm, right? And this can make our code more understandable. We donot have to care whether the user set it wrong or we set it wrong, this is a double-check and not expensive. Awesome!

Reply:
this will be duplicate to me, so I prefer to put it to the else, the logic will be if user specify a grant type, we will validate it, if user doesn't specify one, we will infer one default value for it. There's no need to do the double-check.

Reply:
Not concerning who specify this value, we decouple the validation and assignment.  And we donot have to care where this value is from, we use the uniformed flow. Is that sound better?

Reply:
shift the order of the two case in if...else...